pallets/treasury-buyout-extension/src/lib.rs

```diff
@@ -0,0 +1,354 @@
+				exchange_amount.saturated_into::<u128>(),
+				exchange_asset_price.into_inner(),
+				basic_asset_price_with_fee.into_inner(),
+			)
+			.map(|b| b.try_into().ok())
+			.flatten()
+			.ok_or(ArithmeticError::Overflow.into());
+
+			buyout_amount
+		}
+
+		fn split_to_buyout_and_exchange(
+			asset: CurrencyIdOf<T>,
+			amount: Amount<BalanceOf<T>>,
+		) -> Result<(BalanceOf<T>, BalanceOf<T>), DispatchError> {
+			match amount {
+				Amount::Buyout(buyout_amount) => {
+					let exchange_amount = Self::calc_amount_to_exchange(asset, buyout_amount)?;
+					Ok((buyout_amount, exchange_amount))
+				},
+				Amount::Exchange(exchange_amount) => {
+					let buyout_amount = Self::calc_buyout_amount(asset, exchange_amount)?;
+					Ok((buyout_amount, exchange_amount))
+				},
+			}
+		}
+
+		fn do_buyout(
+			who: AccountIdOf<T>,
+			asset: CurrencyIdOf<T>,
+			amount: Amount<BalanceOf<T>>,
+		) -> DispatchResult {
+			Self::ensure_not_native_buyout(&asset)?;
+			let basic_asset = <T as orml_currencies::Config>::GetNativeCurrencyId::get();
+			let (buyout_amount, exchange_amount) =
+				Self::split_to_buyout_and_exchange(asset, amount)?;
+			Self::ensure_buyout_limit_not_exceeded(&who, buyout_amount)?;
+			let treasury_account_id = T::TreasuryAccount::get();
+
+			// Start exchanging
+			// Check for exchanging zero values and same accounts
+			if exchange_amount.is_zero() && buyout_amount.is_zero() || who == treasury_account_id {
+				return Ok(())
+			}
+
+			// Check both balances before transfer
+			let user_balance = T::Currency::free_balance(asset, &who);
+			let treasury_balance = T::Currency::free_balance(basic_asset, &treasury_account_id);
+
+			if user_balance < exchange_amount {
+				return Err(Error::<T>::InsufficientAccountBalance.into())
+			}
+			if treasury_balance < buyout_amount {
+				return Err(Error::<T>::InsufficientTreasuryBalance.into())
+			}
+
+			// Transfer from user account to treasury then viceversa
+			T::Currency::transfer(asset, &who, &treasury_account_id, exchange_amount)
+				.map_err(|_| Error::<T>::ExchangeFailure)?;
+			T::Currency::transfer(basic_asset, &treasury_account_id, &who, buyout_amount)
+				.map_err(|_| Error::<T>::ExchangeFailure)?;
+
+			Self::update_buyouts(&who, buyout_amount);
+			Self::deposit_event(Event::<T>::Buyout { who, buyout_amount, asset, exchange_amount });
+
+			Ok(())
+		}
+
+		fn multiply_by_rational(
+			a: impl Into<u128>,
+			b: impl Into<u128>,
+			c: impl Into<u128>,
+		) -> Option<u128> {
+			// return a * b / c
+			sp_runtime::helpers_128bit::multiply_by_rational_with_rounding(
+				a.into(),
+				b.into(),
+				c.into(),
+				Rounding::NearestPrefDown,
+			)
+		}
+
+        // Use NoPrice error here maybe
+		fn fetch_prices(
+			assets: (&CurrencyIdOf<T>, &CurrencyIdOf<T>),
+		) -> Result<(FixedU128, FixedU128), DispatchError> {
+			let basic_asset_price: FixedU128 =
+				T::PriceGetter::get_price::<FixedU128>(*assets.0)?.into();
```

Comment:
Here it will [need](https://github.com/pendulum-chain/pendulum/pull/396/files#diff-b2aa68998afa0af43db27d3ee43b0298c6d614ee556a1b60969a00a9fe514696R247) to return the price of the native asset but with the fee, is this something we will add to directly to the implementation of the `PriceGetter`? 

Otherwise we could use the `SellerFee` getter to calculate (+one and multiply) the value with fee here. I see that equilibrium does it [here](https://github.com/eq-lab/equilibrium/blob/ddab6da9fdbca80c7988b21c3eff96712d576f64/pallets/eq-treasury/src/lib.rs#L299)

Reply:
True, re-added that now since I lost that part while refactoring